File: Backend/routers/text.py
from fastapi import APIRouter
from typing import Dict, Any

from schemas import TextProcessRequest, TextCompareRequest
from services.text_service import TextService
from core import ValidationException

router = APIRouter(prefix="/api/text", tags=["text"])


@router.post("/process", summary="处理文本", description="支持JSON格式化、Base64编码解码、URL编码解码")
async def process_text(request: TextProcessRequest) -> Dict[str, Any]:
    """
    处理文本，根据指定的操作类型进行相应处理
    
    支持的操作类型：
    - json_format: JSON格式化
    - base64_encode: Base64编码
    - base64_decode: Base64解码
    - url_encode: URL编码
    - url_decode: URL解码
    """
    try:
        result = TextService.process_text(request.action.value, request.text)
        return result
    except ValueError as e:
        raise ValidationException(str(e))


@router.post("/compare", summary="比较文本差异", description="对比两个文本的差异并返回详细结果")
async def compare_text(request: TextCompareRequest) -> Dict[str, Any]:
    """
    比较两个文本的差异
    
    返回差异列表和摘要信息
    """
    result = TextService.compare_text(request.text1, request.text2)
    return result

File: Backup/tools_project_rust/src/services/tools/text.rs
use crate::core::error::{bad_request, AppResult};
use crate::models::{CompareSummary, LineDifference, TextAction};
use base64::prelude::*;

pub fn process_text(action: &TextAction, text: &str) -> AppResult<String> {
    match action {
        TextAction::JsonFormat => {
            let data: serde_json::Value = serde_json::from_str(text)
                .map_err(|e| bad_request(format!("无效的JSON格式: {}", e)))?;
            Ok(serde_json::to_string_pretty(&data).unwrap_or_default())
        }
        TextAction::Base64Encode => Ok(BASE64_STANDARD.encode(text.as_bytes())),
        TextAction::Base64Decode => {
            let bytes = BASE64_STANDARD
                .decode(text)
                .map_err(|e| bad_request(format!("Base64解码失败: {}", e)))?;
            String::from_utf8(bytes).map_err(|e| bad_request(format!("Base64解码失败: {}", e)))
        }
        TextAction::UrlEncode => Ok(urlencoding::encode(text).to_string()),
        TextAction::UrlDecode => {
            urlencoding::decode(text)
                .map(|s| s.to_string())
                .map_err(|e| bad_request(format!("URL解码失败: {}", e)))
        }
    }
}

pub fn compare_text(text1: &str, text2: &str) -> (Vec<LineDifference>, CompareSummary) {
    let lines1: Vec<&str> = text1.lines().collect();
    let lines2: Vec<&str> = text2.lines().collect();
    let max_lines = lines1.len().max(lines2.len());
    let mut differences = Vec::new();

    for i in 0..max_lines {
        let line1 = lines1.get(i).unwrap_or(&"").to_string();
        let line2 = lines2.get(i).unwrap_or(&"").to_string();

        if line1 != line2 {
            let diff_type = if line1.is_empty() {
                "added"
            } else if line2.is_empty() {
                "removed"
            } else {
                "modified"
            };
            differences.push(LineDifference {
                line: i + 1,
                text1: line1,
                text2: line2,
                diff_type: diff_type.to_string(),
            });
        }
    }

    let summary = CompareSummary {
        total_lines: max_lines,
        different_lines: differences.len(),
        identical: differences.is_empty(),
    };

    (differences, summary)
}
